### PR TITLE
3.0 proc type mapping

### PR DESCRIPTION
**Pull request by boggle**

### CODE REVIEW

File: integrationtests/src/test/java/org/neo4j/procedure/ProcedureIT.java [hunk 1]
```diff
@@ -280,6 +281,81 @@ public void procedureMarkedForWritingCanCallOtherWritingProcedure() throws Throw
...
+            Result res = db.execute( "CALL org.neo4j.procedure.squareLong", map( "value", 4L ) );
+
+            // Then
+            assertThat( res.next(), equalTo( map( "someVal", 16L ) ) );
+            assertFalse( res.hasNext() );
+        }
+    }
+
+    @Test
+    public void foo() throws Throwable
```
Comment (pontusmelke) [inline integrationtests/src/test/java/org/neo4j/procedure/ProcedureIT.java]: non-descriptive name :)

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/coerce.scala [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import org.neo4j.cypher.internal.frontend.v3_0.CypherTypeException
+import org.neo4j.cypher.internal.frontend.v3_0.symbols._
+import org.neo4j.graphdb.{Node, Path, Relationship}
+
+object coerce {
+
+  def apply(value: Any, typ: CypherType)(implicit context: QueryContext): Any = {
+    val result = try {
+      if (value == null) null
+      else typ match {
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/coerce.scala]: Could move the `try` in one step; `val result = if (value == null) null else try { ...`

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/procs/coerceToTypeIfNeeded.scala [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+case class coerceToTypeIfNeeded(types: ASTAnnotationMap[Expression, ExpressionTypeInfo])
+  extends ((Expression, CypherType) => Expression) {
+
+  def apply(expr: Expression, expected: CypherType) = {
+    val typeInfo = types(expr)
+    val actual = typeInfo.actual
+    if (actual.iterator.forall(expected.isAssignableFrom)) {
+      expr
+    } else {
+      val optCypherType =
```
Comment (systay) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/procs/coerceToTypeIfNeeded.scala]: This expression is hard to grokk. Could it be disentangled a bit?

**Comment by boggle** [comment]
I've addressed all feedback and rebased; let's see this go green

**Pull request merged**